File: typescript/jstools/tools.ts
export module convert {

    /**
    * Decodes a Base64 encododed timestamp series.
    * 
    * @param encodedStr Encoded timestamps
    * @return Array of Date instances
    */
    export function decodeTimeStamps(encodedStr: string): Date[] {
        if (!encodedStr) return [];

        var stream = atob(encodedStr);
        if (stream.length % 2 > 0) throw 'InvalidInput';

        var readBlock = (stream: string, position: number = 0): number => {
            var bytes: number = 0;
            for (var j = 0; j < 4; j++)
                bytes |= (stream.charCodeAt(position + j) << (j * 8));
            return bytes >>> 0;
        };

        var head = readBlock(stream);
        var beginDate = new Date(new Date(2000, 0, 1).getTime() + (head * 1000));

        var timeStamps: Date[] = new Array();
        timeStamps.push(beginDate);

        var getDiffs = (bytes: number): number[]=> {
            var flag = bytes >>> 30,
                blocks = flag + 1,
                blocksize = (30 / blocks) << 0,
                mask = Math.pow(2, blocksize) - 1;

            var result: number[] = new Array(blocks);
            for (var i = 0; i < blocks; i++)
                result[(blocks - 1) - i] = (bytes >>> (i * blocksize)) & mask;
            return result;
        };

        var prevDate = beginDate,
            i = 4;
        while (i < stream.length) {
            var bytes = readBlock(stream, i);
            var diffs = getDiffs(bytes);
            for (var k = 0; k < diffs.length; k++)
                timeStamps.push((prevDate = new Date(prevDate.getTime() + (diffs[k] * 1000))));
            i += 4;
        }

        return timeStamps;
    }

    /**
    * Encodes timestamp series into a Base64 string.
    * 
    * @param items Timestamps to encode
    * @return Base64 encoded timestamps
    */
    export function encodeTimeStamps(items: Date[]): string {
        if (items.length == 0) return null;

        var diffs: number[] = [];

        var i = 1,
            buckets = new Array<number>(4);
        while (i < items.length) {
            var bucketPtr = 0,
                flag = 4;

            var diff = ((items[i].getTime() - items[i - 1].getTime()) / 1000) << 0;
            do {
                if (diff < (1 << (30 / flag))) {
                    buckets[bucketPtr] = diff;
                    bucketPtr++;
                    i++;

                    if (i < items.length) {
                        if (items[i - 1] > items[i])
                            throw 'NegativeDifference';
                        else
                            diff = ((items[i].getTime() - items[i - 1].getTime()) / 1000) << 0;
                    }
                }
                else {
                    flag--;
                }
            }
            while (flag > 0 && bucketPtr < flag && i < items.length);

            if (flag == 0)
                throw 'FlagOverflow';

            var block = 0;
            for (var j = 0; j < bucketPtr; j++)
                block |= buckets[bucketPtr - 1 - j] << (((30 / bucketPtr) << 0) * j);

            block = (block | (bucketPtr - 1) << 30) >>> 0;

            diffs.push(block);
        }

        var getBytes = (num: number): number[]=> {
            var arr = [
                (num >>> 0) & 0x000000ff,
                (num >>> 8) & 0x000000ff,
                (num >>> 16) & 0x000000ff,
                (num >>> 24) & 0x000000ff
            ];
            return arr;
        };

        var startDay = ((items[0].getTime() - new Date(2000, 0, 1, 0, 0, 0, 0).getTime()) / 1000) << 0,
            buffer = getBytes(startDay);

        diffs.forEach(d => buffer.push.apply(buffer, getBytes(d)));

        var result = '';
        buffer.forEach(b => result += String.fromCharCode(b));

        return window.btoa(result);
    }
}

File: csharp/tools/Tools.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace rblt.Tools
{
    public static class Convert
    {
        #region Converters

        /// <summary>
        /// Encodes and shortens a series of timestamps into BASE64 string.
        /// </summary>
        /// <param name="timeStamps">A series of timestamp instances of type ITimeStampUTC.</param>
        /// <returns>The corresponding BASE64 encoded string.</returns>
        /// <remarks>
        /// The timestamps are encoded in the following way, and the encoded data consists of 32 bits blocks:
        /// 
        ///     - first block contains the first timestamp (passed seconds since 2000-01-01.)
        ///     - any subsequent block contains the difference in seconds from its previous timestamp
        ///         -- a block can hold up to 4 subsequent differences (slots), according to the greatness of
        ///         the values:
        ///         
        ///                  (*) flag      maxvalue     count
        ///                         0        2^30-1         1
        ///                         1        2^15-1         2
        ///                         2        2^10-1         3
        ///                         3         2^7-1         4
        ///                         
        ///          (*) When a block is encoded, the first two bits will hold the value of the block's flag,
        ///          aka. the number of differences minus one.
        ///          -- at the last block, if more slots could be used, than the number of differences left,
        ///          the flag is set according to the remaining count.
        /// </remarks>
        public static string EncodeTimeStamps(IEnumerable<DateTime> timeStamps)
        {
            var items = timeStamps.ToArray();
            if (items.Length == 0) return null;

            List<uint> diffs = new List<uint>();
            int i = 1;
            uint[] buckets = new uint[4];
            while (i < items.Length)
            {
                int bucketPtr = 0,
                    flag = 4;

                uint diff = (uint)(items[i] - items[i - 1]).TotalSeconds;
                do
                {
                    if (diff < (1 << (30 / flag)))
                    {
                        buckets[bucketPtr] = diff;
                        bucketPtr++;
                        i++;

                        if (i < items.Length)
                        {
                            if (items[i - 1] > items[i])
                                throw new ArgumentOutOfRangeException("Negative difference. Timestamp sequence is not monotonic.", (Exception)null);
                            else
                                diff = (uint)(items[i] - items[i - 1]).TotalSeconds;
                        }
                    }
                    else
                    {
                        flag--;
                    }
                }
                while (flag > 0 && bucketPtr < flag && i < items.Length);

                if (flag == 0)
                    throw new OverflowException();

                uint block = 0;
                for (int j = 0; j < bucketPtr; j++)
                    block |= buckets[bucketPtr - 1 - j] << ((30 / bucketPtr) * j);

                block |= (uint)(bucketPtr - 1) << 30;

                diffs.Add(block);
            }

            uint startDay = (uint)(items[0] - new DateTime(2000, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc)).TotalSeconds;
            byte[] buffer = BitConverter.GetBytes(startDay);

            buffer = buffer.Concat(diffs.SelectMany(b => BitConverter.GetBytes(b))).ToArray();

            return System.Convert.ToBase64String(buffer);
        }

        /// <summary>
        /// Decodes the string representation of a series of timestamps.
        /// </summary>
        /// <param name="encodedTimeStamps">The encoded timestamps.</param>
        /// <returns>The corresponding array of DateTime instances.</returns>
        public static DateTime[] DecodeTimeStamps(string encodedTimeStamps)
        {
            if (String.IsNullOrEmpty(encodedTimeStamps)) return new DateTime[0];

            using (var stream = new MemoryStream(System.Convert.FromBase64String(encodedTimeStamps), false))
            {
                if (stream.Length < 2) throw new FormatException("Invalid byte length.");
                stream.Position = 0;

                using (BinaryReader reader = new BinaryReader(stream))
                {
                    var head = reader.ReadUInt32();
                    var beginDate = new DateTime(2000, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc).AddSeconds(head);

                    List<DateTime> timeStamps = new List<DateTime>();

                    Func<uint, uint[]> getDiffs = (uint bytes) =>
                    {
                        uint flag = bytes >> 30,
                            blocks = flag + 1,
                            blocksize = 30 / blocks,
                            mask = (uint)(1 << ((int)blocksize)) - 1;

                        var result = new uint[blocks];
                        for (int i = 0; i < blocks; i++)
                            result[(blocks - 1) - i] = (bytes >> (int)(i * blocksize)) & mask;

                        return result;
                    };

                    var prevDate = beginDate;
                    while (reader.BaseStream.Position < reader.BaseStream.Length)
                    {
                        timeStamps.AddRange(getDiffs(reader.ReadUInt32()).Select(diff => (prevDate = prevDate.AddSeconds(diff))));
                    }

                    timeStamps.Insert(0, beginDate);
                    return timeStamps.ToArray();
                }
            }
        }

        #endregion
    }
}
